src/mlpack/core/optimizers/cmaes/cmaes_impl.hpp

```diff
@@ -0,0 +1,735 @@
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_OPTIMIZERS_CMAES_CMAES_IMPL_HPP
+#define MLPACK_CORE_OPTIMIZERS_CMAES_CMAES_IMPL_HPP
+
+#include <mlpack/core.hpp>
+#include <cassert>
+#include <cmath>
+#include <cstdio>
+#include <cstring>
+#include <fstream>
+#include <limits>
+#include <iomanip>
+#include <sstream>
+#include <stdexcept>
+#include <string>
+#include <armadillo>
+#include <iostream>
+#include <cfloat>
+
+#include "cmaes.hpp"
+#include "random.hpp"
+
+namespace mlpack {
+namespace optimization {
+
+  template<typename funcType>
+  CMAES<funcType>::CMAES(funcType& function,
+arma::mat& start, arma::mat& stdDivs, 
+double iters, double evalDiff)
+      :
+        function(function),
+        N(-1),
+        typicalXcase(false),
+        stopMaxFunEvals(-1),
+        facmaxeval(1.0),
+        stopTolFunHist(1e-14),
+        stopTolX(0),
+        stopTolUpXFactor(1e3),
+        lambda(-1),
+        mu(-1),
+        mucov(-1),
+        mueff(-1),
+        damps(-1),
+        cs(-1),
+        ccumcov(-1),
+        ccov(-1),
+        facupdateCmode(1),
+        weightMode(UNINITIALIZED_WEIGHTS)
+  {
+    stStopFitness.flg = false;
+    stStopFitness.val = -std::numeric_limits<double>::max();
+    updateCmode.modulo = -1;
+    updateCmode.maxtime = -1;
+
+     N = function.NumFunctions();
+    if ( N <= 0)
+      throw std::runtime_error("Problem dimension N undefined.");
+    bool startP  = true;
+    bool initDev = true;
+
+    for (int i=0; i<N; i++)
+    {
+     if (start[i]   < 1.0e-200) startP  = false;
+     if (stdDivs[i] < 1.0e-200) initDev = false;
+    }
+
+    if (evalDiff =! 1e-14) stopTolFun = evalDiff;
+    else
+      stopTolFun = 1e-14;
+
+if (!startP)
+Log::Warn << " WARNING: initial start point undefined." <<
+"Please specify if incorrect results detected."
+<< "DEFAULT = 0.5...0.5." << std::endl;
+
+if (!initDev)
+Log::Warn << "WARNING: initialStandardDeviations undefined."
+<< " Please specify if incorrect results detected. DEFAULT = 0.3...0.3."
+<< std::endl;
+
+    if (weightMode == UNINITIALIZED_WEIGHTS)
+      weightMode = LOG_WEIGHTS;
+
+    diagonalCov = 0; // default is 0, but this might change in future
+
+      xstart.set_size(N);
+      if (startP)
+      {
+        for (int i = 0; i < N; ++i) xstart[i] = start[i];
+      }
+     else
+      {
+        typicalXcase = true;
+        for (int i = 0; i < N; i++) xstart[i] = 0.5;
+      }
+
+    rgInitialStds.set_size(N);
+    if (initDev)
+      {
+        for (int i = 0; i < N; ++i) rgInitialStds[i] = stdDivs[i];
+      }
+      else
+      {
+        for (int i = 0; i < N; ++i) rgInitialStds[i] = double(0.3);
+      }
+
+    if (lambda < 2)
+      lambda = 4 + (int) (3.0*log((double) N));
+    if (mu <= 0)
+      mu = lambda / 2;
+
+      weights.set_size(mu);
```

Comment:
I think we should use the policy design pattern here, to abstract the weight initialization:

```
template<typename funcType, WeightInitPolicy = LinearWeights>
class CMAES
{
CMAES(const WeightInitPolicy& initPolicy = WeightInitPolicy())
{

    //for (int i = 0; i < mu; ++i) weights[i] = mu - i;
    initPolicy.Initialize(weights);
}


}

class LinearWeights
{
  LinearWeights(...) {}
  static void Initialize(arma::mat& weights)
  {
    arma::Col<size_t> spaces = arma::linspace<arma::Col<size_t>>(0,
        (weights.n_elem - 1), weights.n_elem
     weights.fill(mu);
     weights - spaces;
  }
}
```